scanpy/tools/_score_genes.py

```diff
@@ -155,12 +155,24 @@ def score_genes(
     control_genes = pd.Index([], dtype="string")
 
     # now pick `ctrl_size` genes from every cut
-    for cut in np.unique(obs_cut.loc[gene_list]):
-        r_genes: pd.Index[str] = obs_cut[obs_cut == cut].index
+    for cut in obs_cut.loc[gene_list].unique():
+        r_genes: pd.Index[str] = obs_cut[
+            (obs_cut == cut) & ~obs_cut.index.isin(gene_list)
+        ].index
+        if len(r_genes) == 0:
+            logg.warning(
+                f"No control genes for this cut. You might want to increase "
+                f"ctrl_size={ctrl_size} to sample from more control genes."
+            )
         if ctrl_size < len(r_genes):
             r_genes = r_genes.to_series().sample(ctrl_size).index
         control_genes = control_genes.union(r_genes.difference(gene_list))
 
+    assert len(control_genes) > 0, "No control genes found."
+    assert not any(
+        g in control_genes for g in gene_list
+    ), "Genes are in both gene_list and control_genes."
```

Comment:
`assert` is only for internal checks. If one of these conditions can be triggered by user supplied data (which seems to be the case in the test you added), you shouldn’t use `assert`.

Instead, you should use `ValueError` (or `TypeError`) if a faulty parameter is passed in, or `RuntimeError` if something failed that shouldn’t (i.e. something neither the user nor we control happened)